File: typescript/createToken/create.buy.token.pumpfun.test.ts
import { Keypair, VersionedTransaction } from "@solana/web3.js";
import { bs58 } from "@project-serum/anchor/dist/cjs/utils/bytes";
import { configDotenv } from "dotenv";
import fs from "fs";
configDotenv();

//Create pumpfun token and first buy

const pk = process.env.PRIVATE_KEY;
const test = async () => {
  try {
    const private_key = pk || "";
    const wallet = Keypair.fromSecretKey(bs58.decode(private_key));
    console.log("wallet:", wallet.publicKey.toBase58());

    // Upload metadata to IPFS
    const image = await fs.openAsBlob("./zz.png");
    const formData = new FormData();
    formData.append("file", image); // Image file
    formData.append("name", "test");
    formData.append("symbol", "TEST");
    formData.append(
      "description",
      "This is an example token created via SolanaPortal.AI"
    );
    formData.append("twitter", "https://x.com/solana");
    formData.append("telegram", "https://x.com/solana");
    formData.append("website", "https://solanaportal.ai");
    formData.append("showName", "true");

    // Create IPFS metadata storage
    const metadataResponse = await fetch("https://pump.fun/api/ipfs", {
      method: "POST",
      body: formData,
    });
    if (metadataResponse.status !== 200) {
      throw new Error("Failed to create IPFS metadata storage");
    }
    const metadataResponseJSON = await metadataResponse.json();

    const param = {
      name: metadataResponseJSON.metadata.name,
      symbol: metadataResponseJSON.metadata.symbol,
      metadataUri: metadataResponseJSON.metadataUri,
      wallet_address: wallet.publicKey.toBase58(), // Your wallet public key
      amount: 0.00001, // amount of SOL or tokens
      //optional
      // slippage: 100, // percent slippage allowed
      // tip: 0.00005, // priority fee
      // type: "jito", // "jito" or "bloxroute"
    };
    const url = "https://api.solanaportal.io/api/create/token/pumpfun";
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(param),
    });
    // return;
    if (response.status === 200) {
      // Successfully generated transaction
      const data = await response.json();
      const txnBuffer = Buffer.from(data, "base64");
      const txn = VersionedTransaction.deserialize(txnBuffer);
      txn.sign([wallet]);
      const signedTxnBuffer = bs58.encode(txn.serialize());
      const jitoResponse = await fetch(
        `https://tokyo.mainnet.block-engine.jito.wtf/api/v1/transactions`,
        {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            jsonrpc: "2.0",
            id: 1,
            method: "sendTransaction",
            params: [signedTxnBuffer],
          }),
        }
      );
      if (jitoResponse.status === 200) {
        const signature = (await jitoResponse.json()).result;
        console.log("- txn succeed", "https://solscan.io/tx/" + signature);
      } else {
        console.log("- txn failed, please check the parameters", data);
      }
    } else {
      console.log(response.statusText); // log error
    }
  } catch (e: any) {
    console.error(e.message);
  }
};

test();

File: python/createToken/create.buy.token.pumpfun.test.py
#!/usr/bin/env python3
from dotenv import load_dotenv
import os
import base64
import base58
import requests
from solana.keypair import Keypair
from solana.transaction import VersionedTransaction

def main():
    # 1) Load your PRIVATE_KEY from .env
    load_dotenv()
    private_key_b58 = os.getenv("PRIVATE_KEY")
    if not private_key_b58:
        raise EnvironmentError("Missing PRIVATE_KEY in environment")

    # 2) Build your Solana wallet Keypair
    secret = base58.b58decode(private_key_b58)
    wallet = Keypair.from_secret_key(secret)
    print(f"wallet: {wallet.public_key}")

    # 3) Upload metadata + image to IPFS (pump.fun)
    ipfs_url = "https://pump.fun/api/ipfs"
    files = {"file": open("./zz.png", "rb")}
    data = {
        "name": "test",
        "symbol": "TEST",
        "description": "This is an example token created via SolanaPortal.AI",
        "twitter": "https://x.com/solana",
        "telegram": "https://x.com/solana",
        "website": "https://solanaportal.ai",
        "showName": "true",
    }

    ipfs_resp = requests.post(ipfs_url, files=files, data=data)
    if ipfs_resp.status_code != 200:
        print("Failed to create IPFS metadata storage:", ipfs_resp.text)
        return

    ipfs_json = ipfs_resp.json()
    # Expecting fields: metadata.name, metadata.symbol, metadataUri
    metadata       = ipfs_json["metadata"]
    metadata_uri   = ipfs_json["metadataUri"]

    # 4) Build the create-token/pumpfun payload
    create_url = "https://api.solanaportal.io/api/create/token/pumpfun"
    param = {
        "name":         metadata["name"],
        "symbol":       metadata["symbol"],
        "metadataUri":  metadata_uri,
        "wallet_address": str(wallet.public_key),
        "amount":       0.00001,
        # optional fields:
        # "slippage": 100,
        # "tip": 0.00005,
        # "type": "jito",
    }

    create_resp = requests.post(create_url, json=param)
    if create_resp.status_code != 200:
        print("Error generating create-token txn:", create_resp.text)
        return

    # 5) Deserialize, sign, and re-serialize the returned transaction
    txn_b64   = create_resp.json()
    txn_bytes = base64.b64decode(txn_b64)
    txn       = VersionedTransaction.deserialize(txn_bytes)
    txn.sign([wallet])
    signed_b58 = base58.b58encode(txn.serialize()).decode()

    # 6) Send the signed transaction via Jito RPC
    jito_url = "https://tokyo.mainnet.block-engine.jito.wtf/api/v1/transactions"
    jito_payload = {
        "jsonrpc": "2.0",
        "id":      1,
        "method":  "sendTransaction",
        "params":  [signed_b58],
    }
    jito_resp = requests.post(jito_url, json=jito_payload)
    if jito_resp.status_code == 200:
        signature = jito_resp.json().get("result")
        print("- txn succeed:", f"https://solscan.io/tx/{signature}")
    else:
        print("- txn failed, please check the parameters:", jito_resp.text)

if __name__ == "__main__":
    main()
